docs/HYPERVISORS.md

```diff
@@ -0,0 +1,85 @@
+
+This mapping of Tasks to resources creates an interesting architecture where things like virtualization capabilities can be consumed transparently without the user explicitly tweaking the virtualization knobs like it would be the case with systems like libvirt/virsh.
+
+The rest of this document summarizes various implementation details we leverage for providing acceleration and isolation services to tasks.
+
+## Device models
+
+Whenever a task gets isolated into a standalone domain (either for performance or security reasons) domain gets assigned a certain, fixed amount of RAM and can have a number of virtual CPUs that would get multiplexed on top the physical CPUs available on the Edge Node. Additionally, a task may request a subset of PCI devices to be exclusively available to it (which makes them unavailable for any other domains and the host itself). Providing RAM, CPU and direct PCI assignment to a given task is one half of a job of any hypervisor. That's the easy part.
+
+The difficult (or at least much more involved part) is how to present a series of devices (network, disk, GPU, console) to the task in such a way as to allow them to be multiplexed to the actual physical devices available on the host. This is know as providing a "device model" to the domain and it is where the art of virtualization really begins.
+
+All domains get presented with the "device model" that consist of a virtual set of buses and virtual devices attached to those busses. Domains can initiate I/O to any of these busses/devices and that I/O gets routed by the hypervisor to the outside of the domain. Servicing that I/O on the host side can then be done either by:
+
+1. Host Linux kernel directly
+2. A user-space program
+
+Obviously, a user-space program will require more context switches and will always be slower than a corresponding code running directly in the host kernel space. But even then, a user-space program can be thought of as a I/O emulation of "last resort" (if there's no direct support in the kernel for the corresponding bus and/or device). This kind of a user space program is known as a Virtual Machine Monitor (VMM). Qemu is the most featureful VMM that exists today in open source and currently EVE is using it for all the hypervisors it supports. However, depending on how our use cases evolve we may start using more lightweight (or specialized) VMMs in the future, such as:
```

Comment:
I would change this to:

```
Normally, a user-space program will require more context switches and will be slower than corresponding code running directly in the host kernel space.
```

There are user-space-only device drivers and such that eliminate context-switches entirely. Let's try to avoid getting into the debate the next time someone who knows user-space things really well reading our docs.

Also, "slower than a corresponding code" should be "slower than corresponding code" (extra "a")

Reply:
fixed